h2/connection.py

```diff
@@ -860,19 +860,28 @@ def reset_stream(self, stream_id, error_code=0):
         self._reset_streams.add(stream_id)
         del self.streams[stream_id]
 
-    def close_connection(self, error_code=0):
+    def close_connection(self, error_code=0, additional_data=None,
+                         last_stream_id=None):
+
         """
         Close a connection, emitting a GOAWAY frame.
 
         :param error_code: (optional) The error code to send in the GOAWAY
             frame.
+        :param additional_data: (optional) Additional debug data indicating
+            a reason for closing the connection.
+        :param last_stream_id: (optional) The last stream which was processed
+            by the sender. Defaults to `highest_inbound_stream_id`.
         :returns: Nothing
         """
         self.state_machine.process_input(ConnectionInputs.SEND_GOAWAY)
 
-        f = GoAwayFrame(0)
-        f.error_code = error_code
-        f.last_stream_id = self.highest_inbound_stream_id
+        f = GoAwayFrame(
+            stream_id=0,
+            last_stream_id=(last_stream_id or self.highest_inbound_stream_id),
```

Comment:
This doesn't *quite* work as expected: if you explicitly pass `last_stream_id=0`, this will cause hyper-h2 to emit the frame with `last_stream_id = self.highest_inbound_stream_id`. We may need a more complex statement outside the initializer, and then also some tests for this new behaviour (which is arguably unrelated to the test for additional data).